Adding direct central difference method to solid mechanics

TheGreatCid opened the pull request:
This is a follow up to #27991.

## Reason
<!--Why do you need this feature or what is the enhancement?-->
This PR addresses the first half of https://github.com/idaholab/moose/issues/28008

## Design 

The "direct" solution update, is calculated as the following:

$a_n=M^{-1}(F^{ext}-F^{int})$
$v_{n+1/2}=v_{n-1/2}+\frac{\Delta t_(old) + \Delta t}{2} a_n$
$d_{n+1}=d_n + \Delta t v_{n+1/2}$

A new class is added to implement the above formulation: `CentralDifferenceDirect`
`CentralDifferenceDirect` performs the solution update while also retrieving the mass matrix from `MassMatrix` which needs to be present in any input file that uses the new time integration method. 

Additionally, a new boundary condition base, `DirectDirichletBCBase` was added to enforce Dirichlet BC's using this new time integrator. It computes the residual force needed to enforce the BC during the solution update.

TheGreatCid committed:
Added central difference method to solid mechanics that calculates acceleration directly from the residual forces #27883

TheGreatCid committed:
Added BC's for use with direct central difference method. These BC's calcuate the forces needed to enforce the dirichlet BC and add them to the residual forces

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

hugary1995 commented on modules/solid_mechanics/include/timeintegrators/CentralDifferenceDirect.h:
```diff
@@ -0,0 +1,56 @@
+  // Applying pre-set nodal BCs
+  void setPresetBCs();
+
+  void computeADTimeDerivatives(ADReal & ad_u_dot,
+                                const dof_id_type & dof,
+                                ADReal & ad_u_dotdot) const override
+  {
+    ad_u_dotdot.value();
+    ad_u_dot.value();
+    std::cout << dof << std::endl;
+    return mooseError("NOT SUPPORTED");
+  }
```
What's going on here?

TheGreatCid replied:
This is a copy of what was found in `CentralDifference.C` to overcome the override requirement for `ComputeADTimeDerivatives`

hugary1995 replied:
But do you need these three lines?
```
    ad_u_dotdot.value();
    ad_u_dot.value();
    std::cout << dof << std::endl;
```

TheGreatCid replied:
Those are too suppress compiler messages. It's a bit jank but I assumed it was acceptable since it was already implemented.

lindsayad replied:
>This is a copy of what was found in CentralDifference.h to overcome the override requirement for ComputeADTimeDerivatives

Not quite sure what you mean here. There are no errors in the AD routine in `CentralDifference` in upstream MOOSE

TheGreatCid replied:
Hmm. you're right. I'm not sure what I mean either. I can't remember where I got this from. 
What is the correct way to handle this case?

lindsayad replied:
If you don't want to support AD, you can just make this function a one-line `mooseError`. And then you can delete the function parameters so that you don't get unused variable warnings

TheGreatCid replied:
[ab8b9fe](https://github.com/idaholab/moose/pull/28175/commits/ab8b9fe85692cb185dda7f6264efbf1c75a0e597)

hugary1995 commented on modules/solid_mechanics/include/timeintegrators/CentralDifferenceDirect.h:
```diff
@@ -0,0 +1,56 @@
+  }
+
+protected:
+  /**
+   * Helper function that actually does the math for computing the time derivative
+   */
```
```suggestion
```

hugary1995 commented on modules/solid_mechanics/src/timeintegrators/CentralDifferenceDirect.C:
```diff
@@ -0,0 +1,194 @@
+#include "libmesh/sparse_matrix.h"
+#include "DirichletBCBase.h"
+
+registerMooseObject("MooseApp", CentralDifferenceDirect);
```
```suggestion
registerMooseObject("SolidMechanicsApp", CentralDifferenceDirect);
```

hugary1995 commented on modules/solid_mechanics/src/timeintegrators/CentralDifferenceDirect.C:
```diff
@@ -0,0 +1,194 @@
+
+  // Scaling the acceleration
+  auto accel_scaled = accel.clone();
+  accel_scaled->scale((_dt + _dt_old) / 2);
```
This is different from the math you showed in the PR description.

TheGreatCid replied:
Good catch, I need to update the documentation. This is the correct version

hugary1995 replied:
Okay, then this is different from the original paper you referenced. I think this will require some documentation (i.e. why (dt+dt_old)/2 instead of dt).

TheGreatCid replied:
I'm guessing the paper I referenced assumed constant dt. The way I've implemented does not

hugary1995 replied:
Why did you choose (dt+dt_old)/2? That doesn't seem to follow the constant acceleration assumption.

hugary1995 replied:
Check my math, assuming constant acceleration and linear velocity
```
v_{n+1/2} = v_{n-1/2} + dt_old / 2 * a_{n-1} + dt / 2 * a_n
```
Your implementation assumes
```
(dt_old+dt) / 2 * a_n = dt_old / 2 * a_{n-1} + dt / 2 * a_n
```
which is questionable.

TheGreatCid replied:
The first statement should be written as 
`v_{n+1/2} = v_{n-1/2} + dt_old / 2 * a_n + dt / 2 * a_n`

Really this is just defining an effective time step to jump between mid points when the time step is not constant. 
`dt_eff = (dt_{n} + dt_{n+1})/2`
`v_{n+1/2} = v_{n-1/2} +  dt_eff  * a_n`

[Abaqus uses the same method. ](https://classes.engineering.wustl.edu/2009/spring/mase5513/abaqus/docs/v6.6/books/gsx/default.htm?startat=ch03s02.html)

If `t_{n-1} = 1`, `t_{n}=2` and `t_{n+1} = 2.5`, we want `t_{n-1/2}+dt_eff = 2.25 = t_{n+1/2}`
In this case `dt_eff=0.75`, so this would return the correct value for`t_{n+1/2}`

hugary1995 replied:
Okay, I see. So this assumes constant acceleration between midpoints, not between time step begin and end. Your last comment can be a good documentation. Make sure you also reference the Abaqus manual. I'll resolve this conversation once I see the relevant documentation added.

TheGreatCid replied:
[f4073ac](https://github.com/idaholab/moose/pull/28175/commits/f4073ace5c7337485b751e904d6d55d78ed7b604)

I've updated the documentation here

hugary1995 commented on modules/solid_mechanics/src/timeintegrators/CentralDifferenceDirect.C:
```diff
@@ -0,0 +1,194 @@
+
+  // Scale velocity by dt
+  auto vel_scaled = vel.clone();
+  vel_scaled->scale(_dt);
```
This is however consistent with your math...

hugary1995 commented on modules/solid_mechanics/src/timeintegrators/CentralDifferenceDirect.C:
```diff
@@ -0,0 +1,194 @@
+  vel.close();
+  accel.close();
+
+  // Check for convergence by seeing if there is a nan or inf
+  auto sum = _solution_update.sum();
+  converged = std::isfinite(sum);
+
+  // The linear iteration count remains zero
+  _n_linear_iterations = 0;
+  return converged;
```
This is fine with me for now. In the long term, we need a better convergence check for explicit dynamics, or rather a _divergence check_.

Imagine we use what you have here as a convergence check: When the solve didn't not "converge", the framework handles solve failure, restores the solution, and travel back in time to reattempt the solve with a reduced time step. For implicit, that will work -- as we are moving closer to the basin of a traction. But for explicit, that doesn't help, in general.

TheGreatCid replied:
I agree, for now I was simply copying the current way of doing things found in `ExplicitTimeIntegrator.C`

hugary1995 commented on modules/solid_mechanics/src/timeintegrators/CentralDifferenceDirect.C:
```diff
@@ -0,0 +1,194 @@
+  _fe_problem.setUDotDotOldRequested(true);
+}
+
+void
+CentralDifferenceDirect::computeTimeDerivatives()
+{
+  /*
+  Because this is called in NonLinearSystemBase
+  this should not actually compute the time derivatives.
+  Calculating time derivatives here will cause issues for the
+  solution update.
+  */
+  return;
+}
```
Where can I check to see if the second time derivative is implemented correctly according to your math?

hugary1995 replied:
nvmd, found it below.

hugary1995 commented on modules/solid_mechanics/src/timeintegrators/CentralDifferenceDirect.C:
```diff
@@ -0,0 +1,194 @@
+  _fe_problem.time() = _fe_problem.timeOld();
+  _nonlinear_implicit_system->update();
+
+  // Compute the residual
+  _explicit_residual.zero();
+  _fe_problem.computeResidual(
+      *_nonlinear_implicit_system->current_local_solution, _explicit_residual, _nl.number());
```
could you help me understand why this _explicit_residual is Fext-Fint in the equation?

TheGreatCid replied:
It will make more sense once I get some test input files pushed in. I've stalled on this PR while I am at WCCM. 
This is the case because Fint will be the residual returned from your stress divergence only. Fext will return any relevant nodal BCs

TheGreatCid replied:
There will be no inertia kernel in the input file

hugary1995 replied:
> Fint will be the residual returned from your stress divergence only. Fext will return any relevant nodal BCs

That makes sense.

> There will be no inertia kernel in the input file

How do you tag the mass matrix then?

TheGreatCid replied:
I've added a test to the PR. This shows how the mass matrix is tagged

TheGreatCid committed:
Added test for direct central difference

TheGreatCid committed:
Adding additional comments+minor cleaning

TheGreatCid commented:
I've added perhaps the more questionable aspect of this PR, which is classes for Dirichlet BC enforcement. I've introduced a new base class which accounts for the central difference time integration to return the proper force required to enforce the Dir BC. I've added a constant BC and function BC on top of this base class.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2358281/) on 37aadaa wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/28175/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/28175/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 89ffe9d1cf4836fb77572002e59c84f63b572365`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2358410/) on 7dacccd wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28175/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2358425/) on 7dacccd wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Solid mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>89ffe9</th>
      <th colspan=4>#28175 7daccc</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.87%</td>
      <td>84.86%</td>
      <td>-0.02%</td>
      <td>80.34%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>27871</td>
      <td>27965</td>
      <td>+94</td>
      <td>94</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>4967</td>
      <td>4990</td>
      <td>+23</td>
      <td>23</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28175/coverage_diff/solid_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28175/coverage/solid_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28175/coverage/xfem)

</details>

## Warnings

- `solid_mechanics` new line coverage rate 80.34% is less than the suggested 90.0%



This comment will be updated on new commits.

TheGreatCid committed:
Added documentation for direct central difference and direct dir. BCs

TheGreatCid commented:
Not sure if this is an issue, but this implementation only outputs midpoint velocities. So the velocity fields will be $v_{n+1/2}$.

lindsayad commented on modules/solid_mechanics/include/bcs/DirectDirichletBCBase.h:
```diff
@@ -0,0 +1,62 @@
+  /// Whether or not the value is to be preset
+  const bool _preset;
+
+  /// Nonlinear implicit system, if applicable; otherwise, nullptr
+  // NonlinearImplicitSystem * _nonlinear_implicit_system;
+  NonlinearImplicitSystem * _nonlinear_implicit_system;
+
```
```suggestion
```

TheGreatCid replied:
[62bfb7d](https://github.com/idaholab/moose/pull/28175/commits/62bfb7d35c75281000aea0b1f35f3dd6acffa259)

lindsayad commented on modules/solid_mechanics/include/timeintegrators/CentralDifferenceDirect.h:
```diff
@@ -0,0 +1,49 @@
+#include "MooseError.h"
+#include "libmesh/sparse_matrix.h"
+
+class MassMatrixHelper;
```
```suggestion
```

lindsayad commented on modules/solid_mechanics/include/timeintegrators/CentralDifferenceDirect.h:
```diff
@@ -0,0 +1,49 @@
+
+#include "ExplicitTimeIntegrator.h"
+#include "MooseError.h"
+#include "libmesh/sparse_matrix.h"
```
can forward declare

TheGreatCid replied:
I don't understand the reason, could you explain?

hugary1995 replied:
https://stackoverflow.com/questions/9906402/should-one-use-forward-declarations-instead-of-includes-wherever-possible

TheGreatCid replied:
I wasn't aware of this aspect of C++, thanks

TheGreatCid replied:
[d8b86f1](https://github.com/idaholab/moose/pull/28175/commits/d8b86f16fa90d2992962003d08f61c78bf953921)

lindsayad commented on modules/solid_mechanics/src/bcs/DirectDirichletBCBase.C:
```diff
@@ -0,0 +1,66 @@
+void
+DirectDirichletBCBase::initialSetup()
+{
+  _mass_matrix = &_nonlinear_implicit_system->get_system_matrix();
```
this confuses me. In your test you have an extra tagged matrix called "mass" which is what the `MassMatrix` kernel adds into. What is in the system matrix? It seems like it's not the mass. What I would expect to see is the `_mass_matrix` corresponding to the tagged mass matrix, not the system matrix. Else the `_mass_matrix` data member should have some different name

lindsayad replied:
If you still need to access the system matrix in this class, I would do it with `_sys.system().get_matrix("System Matrix")`. In that case you will get a map find error if there is no system matrix which could happen if the system is not an `ImplicitSystem`. Using `_sys.system().get_matrix("System Matrix")` also has the conceptual benefit of working with both `NonlinearImplicitSystem` and `LinearImplicitSystem`.

TheGreatCid replied:
The logic here is that `CentralDifferenceDirect::performExplicitSolve` compute the mass matrix and stores it in `_nonlinear_implicit_system->get_system_matrix()` before the residual is calculated. `DirectDirichletBCBase.C` is retrieving that matrix. 
I don't believe the system matrix is ever written by anything else. This is the same logic currently used in `ActuallyExplicitEuler`.

TheGreatCid replied:
I believe this is resolved here, by tagging the system matrix instead of generating a new tag.
[9ece25d](https://github.com/idaholab/moose/pull/28175/commits/9ece25dbf1c22cb308158e04798e1903c05d45dc)

lindsayad replied:
agreed!

lindsayad commented on modules/solid_mechanics/src/bcs/DirectDirichletBCBase.C:
```diff
@@ -0,0 +1,66 @@
+
+DirectDirichletBCBase::DirectDirichletBCBase(const InputParameters & parameters)
+  : NodalBC(parameters),
+    _preset(false),
+    _nonlinear_implicit_system(dynamic_cast<NonlinearImplicitSystem *>(&_sys.system())),
```
```suggestion
```

TheGreatCid replied:
[62bfb7d](https://github.com/idaholab/moose/pull/28175/commits/62bfb7d35c75281000aea0b1f35f3dd6acffa259)

lindsayad commented on modules/solid_mechanics/src/bcs/DirectDirichletBCBase.C:
```diff
@@ -0,0 +1,66 @@
+
+  // Get lumped component of mass matrix
+  Real masslump = 0;
+  for (int j = 0; j < int(_mass_matrix->n()); j++)
```
```suggestion
  for (const auto j : make_range(_mass_matrix->n()))
```

lindsayad commented on modules/solid_mechanics/src/bcs/DirectDirichletBCBase.C:
```diff
@@ -0,0 +1,66 @@
+DirectDirichletBCBase::computeQpResidual()
+{
+  // Get dof for current var
+  auto dofnum =
```
```suggestion
  const auto dofnum =
```

lindsayad commented on modules/solid_mechanics/src/timeintegrators/CentralDifferenceDirect.C:
```diff
@@ -0,0 +1,161 @@
+void
+CentralDifferenceDirect::solve()
+{
+  // Getting the tagID for the mass matrix
+  auto mass_tag = _sys.subproblem().getMatrixTagID(_mass_matrix);
+
+  // Reset iteration counts
+  _n_nonlinear_iterations = 0;
+  _n_linear_iterations = 0;
+
+  _current_time = _fe_problem.time();
+
+  auto & mass_matrix = _nonlinear_implicit_system->get_system_matrix();
+
+  // Compute the mass matrix
+  if (!_constant_mass || (_constant_mass && _t_step == 1))
+    _fe_problem.computeJacobianTag(
+        *_nonlinear_implicit_system->current_local_solution, mass_matrix, mass_tag);
```
I see. It looks like you are using the mass matrix tag ID, but you are not using the extra matrix that is added for that tag ID. So right now you will have two matrices in your system, one of which is not used. Maybe you should just add the `MassMatrix` contributions directly into the "system" matrix tag instead of creating the extra matrix with `extra_tag_matrices`

TheGreatCid replied:
Hmm. Perhaps I am confused as to how these systems works. My understanding is that the mass matrix is computed using `computeJacobianTag`. The `MassMatrix` tag is used in `computeJacobianTag` to properly calculate the mass matrix from `MassMatrix.C`. 
 It is then stored in `_nonlinear_implicit_system` as that is where `mass_matrix` is referencing. This leaves open the option to use a constant mass.

TheGreatCid replied:
Does the mass matrix tag ID actually generate a matrix, even without the call to `computeJacobianTag`?

TheGreatCid replied:
Oh I see what you are saying. Ignore the previous comments. That make sense.

TheGreatCid replied:
[9ece25d](https://github.com/idaholab/moose/pull/28175/commits/9ece25dbf1c22cb308158e04798e1903c05d45dc)

lindsayad commented on modules/solid_mechanics/test/tests/dynamics/time_integration/direct_central_difference.i:
```diff
@@ -0,0 +1,135 @@
+
+    [TimeIntegrator]
+        type = CentralDifferenceDirect
+        mass_matrix_tag = 'mass'
```
```suggestion
        mass_matrix_tag = 'system'
```

lindsayad commented on modules/solid_mechanics/test/tests/dynamics/time_integration/direct_central_difference.i:
```diff
@@ -0,0 +1,135 @@
+    [massmatrix]
+        type = MassMatrix
+        density = 1
+        matrix_tags = 'mass'
```
```suggestion
        matrix_tags = 'system'
```

TheGreatCid replied:
[9ece25d](https://github.com/idaholab/moose/pull/28175/commits/9ece25dbf1c22cb308158e04798e1903c05d45dc)

lindsayad commented on modules/solid_mechanics/test/tests/dynamics/time_integration/direct_central_difference.i:
```diff
@@ -0,0 +1,135 @@
+    [massmatrix_y]
+        type = MassMatrix
+        density = 1
+        matrix_tags = 'mass'
```
```suggestion
        matrix_tags = 'system'
```

TheGreatCid replied:
[9ece25d](https://github.com/idaholab/moose/pull/28175/commits/9ece25dbf1c22cb308158e04798e1903c05d45dc)

lindsayad commented on modules/solid_mechanics/test/tests/dynamics/time_integration/tests:
```diff
@@ -72,4 +72,13 @@
+    input = 'direct_central_difference.i'
+    exodiff = 'direct_central_difference_out.e'
+    abs_zero = 1e-09
+    requirement = "The solid mechanics app shall support a generalized form of central difference time integration"
```
I think you should be more descriptive than "generalized", communicating what this new time integrator does that the framework integrator does not

TheGreatCid replied:
Fair point, Ill change that

TheGreatCid replied:
[d9b75bb](https://github.com/idaholab/moose/pull/28175/commits/d9b75bb14970b3f38865683f49e6bac2fd62edc1)

TheGreatCid committed:
Adding forward declarations

lindsayad commented on modules/solid_mechanics/include/bcs/DirectDirichletBCBase.h:
```diff
@@ -0,0 +1,62 @@
+
+  void computeValue(NumericVector<Number> & current_solution);
+
+  bool preset() const { return _preset; }
```
It would if your class was derived from `DirichletBCBase` but your class is derived directly from `NodalBC`. Another way to think about it is that since this function is not an overridden virtual it is impossible for it to be called from the framework

TheGreatCid replied:
[ab8b9fe](https://github.com/idaholab/moose/pull/28175/commits/ab8b9fe85692cb185dda7f6264efbf1c75a0e597)

lindsayad commented on modules/solid_mechanics/include/timeintegrators/CentralDifferenceDirect.h:
```diff
@@ -0,0 +1,54 @@
+template <typename T>
+class SparseMatrix;
+}
+class MooseError;
```
Since you actually call the function here in the header you need the definition of `mooseError` available so you actually need to `#include "MooseError.h"`. However, you're "getting lucky" because it's already indirectly included via `#include "ExplicitTimeIntegrator.h"`. I would remove this line or alternatively, move your `computeADTimeDerivatives` implementation to the .C file

TheGreatCid replied:
[ab8b9fe](https://github.com/idaholab/moose/pull/28175/commits/ab8b9fe85692cb185dda7f6264efbf1c75a0e597)

I just added the include statement and removed the forward declaration

lindsayad commented on modules/solid_mechanics/test/tests/dynamics/time_integration/direct_central_difference.i:
```diff
@@ -0,0 +1,135 @@
+# method
+###########################################################
+
+# [Problem]
+#     extra_tag_matrices = 'mass'
+# []
+
```
```suggestion
```

TheGreatCid replied:
[ab8b9fe](https://github.com/idaholab/moose/pull/28175/commits/ab8b9fe85692cb185dda7f6264efbf1c75a0e597)

lindsayad commented on modules/solid_mechanics/test/tests/dynamics/time_integration/tests:
```diff
@@ -72,4 +72,16 @@
+    input = 'direct_central_difference.i'
+    exodiff = 'direct_central_difference_out.e'
+    abs_zero = 1e-09
+    requirement = "The solid mechanics app shall support a form of central difference time integration
```
this is a great requirement, thanks!

lindsayad reviewed (commented):
this is looking quite good to me

TheGreatCid commented:
Should I manually squash all the additional commits, or is the repo set to automatically do that?

lindsayad commented on modules/solid_mechanics/include/timeintegrators/CentralDifferenceDirect.h:
```diff
@@ -38,13 +37,9 @@ class CentralDifferenceDirect : public ExplicitTimeIntegrator
-    ad_u_dotdot.value();
-    ad_u_dot.value();
-    return mooseError("On %d - NOT SUPPORTED", dof);
+    return mooseError("NOT SUPPORTED");
```
no need to `return` here

TheGreatCid replied:
[736c3f4](https://github.com/idaholab/moose/pull/28175/commits/736c3f45bea7fbcebac66ba88f30a5be7e0b0c79)

lindsayad commented:
down to 2 conversations!

TheGreatCid committed:
Renaming CentralDifferenceDirect to DirectCentralDifference for consistency

hugary1995 commented:
I know one of the 2 remaining conversations is mine. I'll have to insist on resolving that one before merging.

TheGreatCid commented:
I renamed `CentralDifferenceDirect` to `DirectCentralDifference` for consistency.

lindsayad commented:
> Should I manually squash all the additional commits, or is the repo set to automatically do that?

the repo does not automatically squash

TheGreatCid commented:
> > Should I manually squash all the additional commits, or is the repo set to automatically do that?
> 
> the repo does not automatically squash

Would you prefer a consolidation of the commits before merging?

GiudGiud commented:
It would be good to squash the "Apply suggestions from code review" with the commits they belong to.
And more importantly, it would be helpful to have messages that are more descriptive than "apply suggestions" or "address reviewer comments"

TheGreatCid committed:
Update direct central difference documentation to explain time step averaging

TheGreatCid commented:
Well it looks like I made some mistakes with my rebasing... I'll need to clean this up.

TheGreatCid commented:
Ok. Back to normal. I'll be more careful rebasing this time.

TheGreatCid commented:
Alright, I've correctly consolidated my commits

lindsayad commented on modules/solid_mechanics/include/bcs/DirectDirichletBCBase.h:
```diff
@@ -0,0 +1,49 @@
+{
+template <typename T>
+class SparseMatrix;
+template <typename T>
+class NumericVector;
+class NonlinearImplicitSystem;
```
```suggestion
```

lindsayad commented on modules/solid_mechanics/include/bcs/DirectDirichletBCBase.h:
```diff
@@ -0,0 +1,49 @@
+
+  virtual void initialSetup() override;
+
+  bool preset() const { return _preset; }
+
```
```suggestion
```

lindsayad commented on modules/solid_mechanics/include/bcs/DirectDirichletBCBase.h:
```diff
@@ -0,0 +1,49 @@
+   */
+  virtual Real computeQpValue() = 0;
+
+  /// Whether or not the value is to be preset
+  const bool _preset;
+
```
```suggestion
```

lindsayad commented on modules/solid_mechanics/include/timeintegrators/DirectCentralDifference.h:
```diff
@@ -0,0 +1,53 @@
+template <typename T>
+class SparseMatrix;
+}
+class MooseError;
```
```suggestion
```

lindsayad commented on modules/solid_mechanics/include/timeintegrators/DirectCentralDifference.h:
```diff
@@ -0,0 +1,53 @@
+class MooseError;
+
+/**
+ * Implements a truly explicit (no nonlinear solve) first-order, forward Euler
+ * time integration scheme.
```
Seems like not quite the right doxygen

TheGreatCid replied:
Fixed this in the latest push

lindsayad commented on modules/solid_mechanics/include/timeintegrators/DirectCentralDifference.h:
```diff
@@ -0,0 +1,53 @@
+
+  virtual bool performExplicitSolve(SparseMatrix<Number> & mass_matrix) override;
+
+  // // Applying pre-set nodal BCs
+  // void setPresetBCs();
+
```
```suggestion
```

lindsayad commented on modules/solid_mechanics/test/tests/dynamics/time_integration/direct_central_difference.i:
```diff
@@ -0,0 +1,135 @@
+# method
+###########################################################
+
+# [Problem]
+#     extra_tag_matrices = 'mass'
+# []
+
```
```suggestion
```

lindsayad reviewed (changes requested):
seems like some changes might have gotten lost in the rebase

TheGreatCid commented:
> seems like some changes might have gotten lost in the rebase

Thank you for catching this.

TheGreatCid committed:
Reapplying suggestions lost in rebase

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

hugary1995 commented on modules/solid_mechanics/doc/content/source/bcs/DirectDirichletBC.md:
```diff
@@ -0,0 +1,21 @@
+
+\begin{equation}
+    F_g = \frac{u_g-u_n}{\Delta t^2} - \frac{v_{n-\frac{1}{2}}}{\Delta t}
+\end{equation}
```
Explain the meaning of the variables.

hugary1995 replied:
Provide an equation for Delta t for changing step size.

TheGreatCid replied:
[a2a74d6](https://github.com/idaholab/moose/pull/28175/commits/a2a74d6e380e6d7fbbd23093bb3390cfa0741a73)

hugary1995 commented on modules/solid_mechanics/doc/content/source/bcs/DirectFunctionDirichletBC.md:
```diff
@@ -0,0 +1,21 @@
+
+\begin{equation}
+    F_g = \frac{u_g-u_n}{\Delta t^2} - \frac{v_{n-\frac{1}{2}}}{\Delta t}
+\end{equation}
```
Explain the meaning of the variables. Point out u_g is the function value.

hugary1995 replied:
Provide an equation for Delta t for charging step size.

hugary1995 commented on modules/solid_mechanics/src/bcs/DirectDirichletBCBase.C:
```diff
@@ -0,0 +1,53 @@
+  // This is the force required to enforce the BC in a central difference scheme
+  Real avg_dt = (_dt + _dt_old) / 2;
+  Real resid =
+      (computeQpValue() - _u[_qp]) / (avg_dt * _dt) - (*_sys.solutionUDotOld())(dofnum) / avg_dt;
```
I'd like to understand the derivation, i.e. why the denominator is (avg_dt * _dt)

TheGreatCid replied:
CD Solution update:
`a_n = M^-1 F` 
`v_(n+1/2) = v_(n-1/2) + avg_dt * a_n`
`d_(n+1) = d_n + dt * v_(n+1/2)`
Here u_(n+1) can be thought of as the boundary value that needs to be enforced, So u_(n+1) = u_g
Then combine the solution update into one equation,

`u_g = u_n + _dt * (v_(n-1/2) + avg_dt M^-1 * F)`
`u_g = u_n + _dt * v_(n-1/2) + _dt * avg_dt * M^-1 * F`
Then solve for F

`F = M*(u_g - u_n  - v_(n-1/2) * _dt)/(_dt * avg_dt)`
 `F = M*[(u_g-u_n) / (_dt*avg_dt) - v_(n-1/2)/avg_dt]`

hugary1995 replied:
Thanks! The derivation makes sense to me. Then shouldn't `(computeQpValue() - _u[_qp])` be `(computeQpValue() - _u_old[_qp])` or am I missing something?

hugary1995 replied:
And more importantly, shouldn't this be whole thing be multiplied by the mass matrix?

TheGreatCid replied:
The mass matrix is multiplied into the residual in the line after the highlighted section here. 

As for your first point, I think you are correct as u_n should equal the old solution at this point in the solve. Though using u_n = _u[_qp] still produces the right answer. There may an assignment of _u[_qp] that I am unaware of that is making the current formulation work by happenstance.

Using `_sys.solutionOld()(dofnum)` returns the same result

hugary1995 replied:
I can understand why _u and _u_old hold the same value at the time of computing the BC residual. I recommend you to request _u_old (and also _u_dot_old) in the constructor. That way we don't rely on undefined (or undocumented) behavior.

TheGreatCid replied:
I know in `_fe_problem` you can set `u_dot_old`  but I don't believe there is a call to to requisition the old solution.

hugary1995 replied:
Use `coupledDotOld`

TheGreatCid replied:
Are you saying to use `coupledDotOld(variable name)` instead of `_sys.solutionOld()(dofnum)` to call the old velocity?
That would involve an extra step as I would need to retrieve the correct variable name

hugary1995 replied:
right... then just use `_var.uDotOld()`, `_var` is a reference to the underlying `MooseVariable`. Similarly use `_var.slnOld()` to get `_u_old`.

TheGreatCid replied:
Why is that better than this?
```
  Real resid = (computeQpValue() - _var.uDotOld()[dofnum]) / (avg_dt * _dt) -
               (_var.slnOld()[dofnum]) / avg_dt;
```

TheGreatCid replied:
Nevermind, that segfaults....

TheGreatCid replied:
Ah but the other method seg faults as well. The `dofnum` parameter is defined based on the total solution vector. I need a separate param for the local index

TheGreatCid replied:
Oh I am overthinking... I can just the _qp index.

TheGreatCid replied:
[8b672f9](https://github.com/idaholab/moose/pull/28175/commits/8b672f9a8c599fc0b8c3018bbd220daabc3c9884)

hugary1995 commented on modules/solid_mechanics/test/tests/dynamics/time_integration/direct_central_difference.i:
```diff
@@ -0,0 +1,131 @@
+    [stress_block]
+        type = ComputeFiniteStrainElasticStress
+    []
+    [density_one]
```
Is the density material property used?

TheGreatCid replied:
It feeds into `WaveSpeed` but that is also not used. I'll remove both and regold.

hugary1995 commented on modules/solid_mechanics/test/tests/dynamics/time_integration/direct_central_difference.i:
```diff
@@ -0,0 +1,131 @@
+    []
+    [massmatrix_y]
+        type = MassMatrix
+        density = 1
```
Please add one more test case with density value that's not one and compare the result against analytical solutions.

TheGreatCid replied:
[23fce96](https://github.com/idaholab/moose/pull/28175/commits/23fce96798e42135935011643212a8d05882c9c3)

The test case with varied time steps uses a non-one density

TheGreatCid replied:
I can add a separate test if you think that would be better

hugary1995 replied:
This works! I assume you also ran the test locally (with a much longer time and more time steps) to make sure the solution matches the analytical solution. Could you do me a favor to attach the figure here (for record keeping)? Thanks!

TheGreatCid replied:
The results from a longer simulation are a bit puzzling. I ran simulations on a single element with `dt=0.01*sin(16*t)+0.015` to oscillate the time steps. This is the same input file as the test added to this PR with the forcing function/dt changed.

When I set `u=t` using a `DirichletFunctionBC` the results look fine, the velocities and the displacement match up.  I plotted the points of the simulations to show the changing time step.

![linearvel](https://github.com/user-attachments/assets/91c1e1d7-c278-441b-8b9d-b237f45e6307)

When I set`u=t*t`, then the results are odd. The solution appears to be correct, however there are oscillations in the mid point velocity that I cannot make sense of. 

![quadvel](https://github.com/user-attachments/assets/f733d1e5-c37b-4d23-a9a5-6502e72095f4)
Here is a zoom in of the phenomenon. 
![quadvel_zoom](https://github.com/user-attachments/assets/469b457d-63dc-4ac5-b38a-d51016c0876f)

The odd thing is that I still get the correct solution despite the mid point velocity appearing incorrect. 

I'm looking into this, but let me know if you have any ideas as to the cause.

hugary1995 replied:
It's difficult for me to visually identify the errors here. Can you perhaps integrate the velocity error in time and show that? I mean something like `\int_t 0.5*(v-v_analytical)^2 dt`. I think some integration error is expected with adaptive time stepping, as long as the error converges uniformly with time step size, it's all good.

TheGreatCid replied:
![Figure_1](https://github.com/user-attachments/assets/1405a8fd-8849-4dda-80eb-f33dac0b440d)
I plotted the average time step(the y-offset of the sin curve for dt) vs. the time integral of the difference between analytical and solution. It does look it is converging ok.

hugary1995 replied:
okay, all good then.

hugary1995 commented on modules/solid_mechanics/test/tests/dynamics/time_integration/direct_central_difference.i:
```diff
@@ -0,0 +1,131 @@
+
+    start_time = 0.0
+    num_steps = 6
+    dt = 0.1
```
Please one more test case with varying time step sizes (given CFL satisfied)

TheGreatCid replied:
[92cea51](https://github.com/idaholab/moose/pull/28175/commits/92cea51147cc62906a06ae97d9587e1f9fee24fa)

Added that test here

hugary1995 commented on modules/solid_mechanics/test/tests/dynamics/time_integration/direct_central_difference.i:
```diff
@@ -0,0 +1,131 @@
+        displacements = 'disp_x disp_y'
+    []
+    [massmatrix]
+        type = MassMatrix
```
Does the mass matrix object accept density defined as a material property? If not I suggest adding one.

TheGreatCid replied:
It does not. It inherits from from `Reaction.h`, and renames the rate param, to be density. It's a bit odd. So most likely the best way to do this would be to add an additional optional parameter that takes in density as a material property. But that would still be a bit weird.

hugary1995 replied:
I would be fine with only keeping the material property parameter. In the input file, you can still assign plain numeric literals to that parameter.

TheGreatCid replied:
In `MassMatrix` the density parameter is defined like this `  params.renameParam("rate", "density", "Optional density for scaling the computed mass.");`. 
Where the original parameter `rate` is a `Real`. 
I'm not sure if I can change this input to a material property without also modifying `Reaction.h`.

hugary1995 replied:
or just create a new object called something like `MatMassMatrix`...

TheGreatCid replied:
That would be the more straight forward option

TheGreatCid replied:
[23fce96](https://github.com/idaholab/moose/pull/28175/commits/23fce96798e42135935011643212a8d05882c9c3)

I took a crack at it here

hugary1995 replied:
It looks good to me. My only complaint is that `MatMassMatrix` and `MassMatrix` have many duplicate code. Could you perhaps factor out a common base class called `MassMatrixBase` and let both derive from it?

hugary1995 replied:
Tagging @lindsayad as it technically belongs to the framework.

GiudGiud replied:
weigh one code duplication against another. how much does it realistically inherit from Reaction.h?

TheGreatCid replied:
[f32f1e7](https://github.com/idaholab/moose/pull/28175/commits/f32f1e751f63f51fb0763df97dd8fae602363d41)

I took a crack at it here

hugary1995 commented:
Some additional suggestions/questions as I forgot to review your DirichletBC changes...

TheGreatCid committed:
Added MatMassMatrix class that computes the mass matrix using material density

TheGreatCid committed:
Added additional test to direct central difference to check the time step averging method for varying time steps

hugary1995 commented:
This also reminds me that we should throw an error if user attempts to use `Predictor` together with central differencing. If you don't already know how to query the warehouse then nevermind, just create an issue so that we don't forget about this in the future.

TheGreatCid commented:
> This also reminds me that we should throw an error if user attempts to use `Predictor` together with central differencing. If you don't already know how to query the warehouse then nevermind, just create an issue so that we don't forget about this in the future.

If would just be this right?

```
  if (_nl.getPredictor())
    mooseError("Predictor not supported for Direct Central Difference!");
```

hugary1995 commented:
Oh if it's that easy just add it I guess. Also add a test to make sure the error can be triggered.

TheGreatCid committed:
Updated direct dir. documentation to include time step averaging and variable definitions

TheGreatCid committed:
Modified how old solution and old velocity are called in DirectDirichletBCBase.h. Additionally removed additional calculation of lumped mass matrix

TheGreatCid commented:
Hmm. The `_nl.getPredictor();` does not seem to work actually. 
It returns a nullptr even if there is a predictor.

hugary1995 commented:
I followed your discussion in #27991. If I understand it correctly, the reason for moving all this into the solid mechanics module is that the `InertialForce` implementation is a bit hacky, which I agree. Now that you've switched to the `MassMatrix` approach which is indeed much cleaner and is receiving first-class framework support, why don't we move the `DirectCentralDifference` stuff into the framework? I don't see anything being very specific to solid mechanics at this point. Or put it another way, this should work for a fluid mechanics problem as well, for example.

hugary1995 commented:
...or even more general than that -- a reaction diffusion problem.

hugary1995 commented:
Other than that, we are down to one final unresolved conversation. I think so far this has been very good work! Consider doing an internship at ANL next summer.

TheGreatCid commented:
That's a good point. Now that we are using the MassMatrix kernel, this should be generalized now. I can draft up some tests with other physics, though I probably won't be able to do that until Monday. 

> Other than that, we are down to one final unresolved conversation. I think so far this has been very good work! Consider doing an internship at ANL next summer.

If I can work on my dissertation while I'm there...

TheGreatCid committed:
Added documentation for MatMassMatrix

TheGreatCid commented:
Gary and I discussed off GitHub to leave this time integrator in solid mechanics instead of moving it to framework.

lindsayad commented:
Don't let him poach you @TheGreatCid or else I will have to request changes again 😆

TheGreatCid committed:
Added a base class for mass matrix from which to derive MassMatrix and MatMassMatrix

TheGreatCid commented:
Well I do agree that there should be a mass matrix base class...

lindsayad commented on framework/include/kernels/MassMatrixBase.h:
```diff
@@ -0,0 +1,27 @@
+/**
+ *  Base class for mass matrix calculation
+ */
+
```
```suggestion
```

lindsayad commented on framework/include/kernels/MatMassMatrix.h:
```diff
@@ -0,0 +1,30 @@
+/**
+ * Calculates a mass matrix using a material property for density.
+ */
+
```
```suggestion
```

lindsayad commented on framework/include/kernels/MatMassMatrix.h:
```diff
@@ -0,0 +1,30 @@
+protected:
+  virtual Real computeQpJacobian() override;
+
+  // The density of the material
```
```suggestion
  /// The density of the material
```

lindsayad commented on modules/solid_mechanics/test/tests/dynamics/time_integration/tests:
```diff
@@ -72,4 +72,29 @@
+    requirement = "The solid mechanics app shall support a form of central difference time integration
+                 that uses a direct calculation of the acceleration from the residual forces to 
+                 perform a solution update. The direct central difference method should properly use 
+                 a time step averaging method to handle non-constant time steps"
```
```suggestion
                 a time step averaging method to handle non-constant time steps."
```

TheGreatCid committed:
Formatting fixes

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

lindsayad commented:
I'm unsubscribing to reduce my notifications. Mention me if I'm needed

TheGreatCid committed:
Consolidated massmatrix type classes into a single class that takes density as a material property and inherits from kernel.h

hugary1995 commented on modules/solid_mechanics/test/tests/dynamics/time_integration/direct_central_difference_varied_dt.i:
```diff
@@ -0,0 +1,136 @@
+        variable = disp_x
+    []
+    [massmatrix_y]
+        type = MatMassMatrix
```
```suggestion
        type = MassMatrix
```

TheGreatCid replied:
Fixed in the latest force push

hugary1995 commented on modules/solid_mechanics/test/tests/dynamics/time_integration/direct_central_difference.i:
```diff
@@ -0,0 +1,126 @@
+        variable = disp_x
+    []
+    [massmatrix_y]
+        type = MatMassMatrix
```
```suggestion
        type = MassMatrix
```

TheGreatCid replied:
Fixed in the latest force push

dschwen commented on framework/src/kernels/MassMatrix.C:
```diff
@@ -27,8 +30,22 @@ MassMatrix::validParams()
+MassMatrix::computeQpResidual()
+{
+  mooseError("Residual should not be calculated for the MassMatrix kernel");
+  return 0;
```
```suggestion
```

TheGreatCid replied:
[890051f](https://github.com/idaholab/moose/pull/28175/commits/890051f10c04ed475c25d51463f6a01c7b2abb42)

dschwen reviewed (approved):
Merging is currently blocked, so if you have a chance and remove that unnecessary return 0 that'd be great. I'll merge as soon as possible.
